### ISSUE TITLE
feat(gp): input-dependent warps (non-stationary transformed GPs)

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/pyrox#204]
## Summary

Let the warp's parameters depend on the input, `G_{φ(x)}`, giving a **non-stationary process from a stationary kernel**. This is the stronger half of the transformed-GP paper and where its accuracy gains actually live.

Two pieces:

1. An optional `X` keyword on the `Likelihood` protocol, threaded through `_ell_numerical`. Backwards compatible; existing likelihoods ignore it.
2. `_apply_warp` in `_warped.py` learns to pass a condition, so `flowjax` conditional bijections and `gauss_flows.Conditioner` work.

The conditioning network itself is **not** written here — `pyrox_nn` and `gauss_flows` already have it.

## Motivation

From the reference implementation's own results table, plain transformed GPs essentially tie SVGP (NLL 2.795 vs 2.798 on `power`); the input-dependent variants are where the improvement is (2.712 / 2.672). [T2] alone delivers the tie.

A single stationary kernel plus an `x`-dependent warp gives non-stationarity with no non-stationary kernel, no deep GP, and no change to the inducing-point structure.

*Verified* — with `b(x)` ramping linearly over a fixed `N(0,1)` base:

| `x` | `b(x)` | `Var[G_x(f)]` | skew |
|---|---|---|---|
| 0.00 | 0.40 | 0.119 | −0.165 |
| 0.50 | 1.20 | 1.968 | −0.639 |
| 1.00 | 2.00 | 18.321 | −1.521 |

A 150× swing in marginal variance and 10× in skewness, from a stationary kernel.

## The protocol change

`Likelihood.log_prob(f, y)` has no `X`, so a conditional warp cannot be reached from `svgp_elbo`. This is the only place existing pyrox code must change.

### `_protocols.py`

```python
    @abstractmethod
    def log_prob(
        self,
        f: Float[Array, " ..."],
        y: Float[Array, " ..."],
        X: Float[Array, " ..."] | None = None,
    ) -> Float[Array, ""]:
        """Conditional ``p(y | f)``, optionally conditioned on inputs.

        ``X`` is supplied only by likelihoods whose parameters are functions
        of the input (see `pyrox_gp.WarpedGaussianLikelihood` with a
        conditional warp). Scalar observation models ignore it.
        """
```

Add `X: ... | None = None` to every concrete likelihood's signature and ignore it. No behaviour changes.

### `_inference.py`

`_ell_numerical` currently vmaps `_ell_one` over `(f_loc, f_var, y)` and calls `lik.log_prob(f, y_n[None])` per point. Add `X` as a fourth vmapped argument:

```python
def _ell_numerical(lik, y, f_loc, f_var, integrator, X=None):
    def _ell_one(mu_n, var_n, y_n, X_n):
        state = GaussianState(
            mean=mu_n[None],
            cov=lx.MatrixLinearOperator(
                var_n[None, None], lx.positive_semidefinite_tag
            ),
        )
        return log_likelihood_expectation(
            lambda f: lik.log_prob(f, y_n[None], X_n),
            state,
            integrator,
        )

    if X is None:
        return jax.vmap(_ell_one, in_axes=(0, 0, 0, None))(
            f_loc, f_var, y, None
        ).sum()
    return jax.vmap(_ell_one)(f_loc, f_var, y, X).sum()
```

`svgp_elbo` already has `X` in scope (`_inference.py:80`); pass it at the call site (line 152). `GaussianLikelihood`'s closed-form branch is untouched.

> **Note on cost.** The ELL evaluates the warp at every quadrature node, so an input-dependent warp costs `order` **network** forward passes per point. At `order=32` that is a 32× multiplier on the conditioner net. `vmap` over nodes with `X` broadcast is the right implementation, but the cost does not go away. Document it.

## `_apply_warp` gains a condition

```python
def _apply_warp(warp, f, X=None):
    """Apply a scalar bijection elementwise, preserving the shape of ``f``.

    For conditional warps, ``X`` is broadcast against ``f``: `svgp_elbo`
    integrates per point, so ``f`` has shape ``(1,)`` and ``X`` is that
    point's ``(D,)`` input, while the predictive path passes ``(N,)`` and
    ``(N, D)``.
    """
    if warp.shape == ():
        fn = lambda v, c: warp.transform(v, c)
    elif warp.shape == (1,):
        fn = lambda v, c: warp.transform(v[None], c)[0]
    else:
        raise ValueError(
            f"Warp event shape must be () or (1,); got {warp.shape}."
        )

    flat = jnp.reshape(f, (-1,))
    if warp.cond_shape is None:
        out = jax.vmap(lambda v: fn(v, None))(flat)
    else:
        if X is None:
            raise ValueError(
                "Conditional warp (cond_shape is not None) requires X. "
                "Pass X through to log_prob, or use an unconditional warp."
            )
        cond = jnp.broadcast_to(
            jnp.reshape(X, (-1, warp.cond_shape[0])),
            (flat.shape[0], warp.cond_shape[0]),
        )
        out = jax.vmap(fn)(flat, cond)
    return jnp.reshape(out, jnp.shape(f))
```

`WarpedGaussianLikelihood.log_prob` and `warped_predictive_moments` both gain the `X` passthrough.

## Usage

The conditioner comes from `gauss_flows` (the `[flows]` extra added in [T2]):

```python
from gauss_flows import Conditioner, MixtureGaussianCDF
from pyrox_gp import WarpedGaussianLikelihood

warp = Conditioner(
    key=key,
    inner=MixtureGaussianCDF(n_components=8, shape=(1,)),
    cond_shape=(X.shape[1],),
    nn_width=32,
    nn_depth=2,
)
lik = WarpedGaussianLikelihood(warp=warp, noise_var=jnp.asarray(0.1))
elbo = svgp_elbo(prior, guide, lik, X, y,
                 integrator=GaussHermiteIntegrator(order=32))
```

`Conditioner` is FiLM-style affine modulation around any unconditional bijection — exactly this use case, and a cleaner factoring than the reference implementation, which bakes two hand-rolled MLPs into the flow class and guards every forward pass with an assertion.

**Do not reimplement the conditioning network.** `pyrox_nn` already has `DenseVariationalDropout` (MC dropout) and `DenseVariational` (mean-field Gaussian) — the two inference schemes the paper uses — plus `FiLM`, `AffineModulation`, and `ConcatConditioner`. If a variational net is used, its KL adds to the ELBO exactly as the reference implementation's flow-KL term does.

## Acceptance criteria

Extend `tests/gp/test_warped.py`; add protocol coverage to `tests/gp/test_likelihoods.py`.

- [ ] **Backwards compatibility.** Every existing likelihood accepts `X=None`
      and every existing test passes unchanged. `GaussianLikelihood`'s
      closed-form ELBO path is byte-identical.
- [ ] **`X=None` is not silently batched.** With an unconditional warp,
      `_ell_numerical` results are identical before and after this change to
      `1e-12`. The `in_axes=(0, 0, 0, None)` branch is the thing being pinned.
- [ ] **Non-stationarity is real.** Hold `f` fixed and vary `X`; assert the
      warped output has non-zero spread across inputs. (Measured `0.197` on the
      prototype.)
- [ ] **Gradients reach the conditioner network.** `eqx.filter_grad` gives
      non-empty finite gradients over the net's leaves — 9 leaves for a
      2-layer `Conditioner`.
- [ ] **Missing `X` raises.** A conditional warp with `X=None` raises
      `ValueError` naming the fix, rather than a shape error from deep inside
      `vmap`.
- [ ] **Shape handling.** Correct for `(f, X)` of `((1,), (D,))` — what
      `svgp_elbo` passes — and `((N,), (N, D))` for the predictive path.
- [ ] **End to end.** One SVI step through `svgp_elbo` with a `Conditioner`
      warp; finite loss, conditioner parameters moved. Mark
      `@pytest.mark.slow`; skip when `gauss_flows` is absent.
- [ ] **Cost is documented, not asserted.** Add the `order`× multiplier to the
      docstring. Do not test timings in CI.

## Design note: why an optional keyword

The alternative — a separate `InputDependentLikelihood` protocol plus a branch in
`svgp_elbo` — was considered and rejected. It is more code, adds a second
dispatch path through the ELBO, and gives nothing an optional keyword does not.
The keyword is invisible to every existing likelihood.

## Out of scope

- The conditioning network itself (`pyrox_nn` / `gauss_flows` own it).
- Variational treatment of the conditioner weights and the resulting KL term.
  Works today via `pyrox_nn.DenseVariational`, but wiring its KL into the ELBO
  is a separate change.
- Threading `X` into the Laplace / EP / posterior-linearization paths. They
  consume `log_prob` and will accept the new keyword, but each has its own
  integration loop; do them on demand.

## Before reporting done

```bash
uv run pytest -q
uv run --group lint ruff check .          # note: '.', not 'packages/'
uv run --group lint ruff format --check .
uv run --group typecheck ty check packages/pyrox-gp/src/pyrox_gp
```

**Reference:** Maroñas, Hamelijnck, Knoblauch & Damoulas (2021), *Transforming Gaussian Processes With Normalizing Flows*, AISTATS, PMLR 130:1081–1089.
